rustworkx-core/src/community/metrics.rs

```diff
@@ -0,0 +1,233 @@
+        Partition {
+            graph,
+            n_subsets: graph.node_count(),
+            node_to_subset: (0..graph.node_count()).collect(),
+        }
+    }
+
+    /// Creates a `Partition` from sets of graph nodes. Checks whether the
+    /// sets actually form a partition of the input graph.
+    pub fn from_subsets(
+        graph: &'g G,
+        subsets: &[HashSet<G::NodeId>],
+    ) -> Result<Partition<'g, G>, NotAPartitionError> {
+        let mut seen = vec![false; graph.node_count()];
+
+        let mut node_to_subset = vec![0; graph.node_count()];
+
+        for (ii, v) in subsets.iter().enumerate() {
+            for &node in v {
+                let idx = graph.to_index(node);
+                if seen[idx] {
+                    // argument `communities` contains a duplicate node
+                    return Err(NotAPartitionError {});
+                }
+                node_to_subset[idx] = ii;
+                seen[idx] = true;
+            }
+        }
+
+        if !seen.iter().all(|&t| t) {
+            return Err(NotAPartitionError {});
+        }
+
+        Ok(Partition::<'g, G> {
+            graph,
+            n_subsets: subsets.len(),
+            node_to_subset,
+        })
+    }
+
+    /// Returns the index of the subset that contains `node`.
+    pub fn subset_idx(&self, node: G::NodeId) -> usize {
+        let idx = self.graph.to_index(node);
+        self.node_to_subset[idx]
+    }
+
+    /// Returns the modularity of the graph with the current partition.
+    pub fn modularity(&self, resolution: f64) -> f64 {
+        let mut internal_weights = vec![0.0; self.n_subsets];
+        let mut outgoing_weights = vec![0.0; self.n_subsets];
+
+        let directed = self.graph.is_directed();
+        let mut incoming_weights = if directed {
+            Some(vec![0.0; self.n_subsets])
+        } else {
+            None
+        };
+
+        for edge in self.graph.edge_references() {
+            let (a, b) = (edge.source(), edge.target());
+            let (c_a, c_b) = (self.subset_idx(a), self.subset_idx(b));
+            let w: f64 = (*edge.weight()).into();
+            if c_a == c_b {
+                internal_weights[c_a] += w;
+            }
+            outgoing_weights[c_a] += w;
+            if let Some(ref mut incoming) = incoming_weights {
+                incoming[c_b] += w;
+            } else {
+                outgoing_weights[c_b] += w;
+            }
+        }
+
+        let sigma_internal: f64 = internal_weights.iter().sum();
+
+        let sigma_total_squared: f64 = if let Some(incoming) = incoming_weights {
+            incoming
+                .iter()
+                .zip(outgoing_weights.iter())
+                .map(|(&x, &y)| x * y)
+                .sum()
+        } else {
+            outgoing_weights.iter().map(|&x| x * x).sum::<f64>() / 4.0
+        };
+
+        let m: f64 = total_edge_weight(self.graph);
+        sigma_internal / m - resolution * sigma_total_squared / (m * m)
```

Comment:
We might want to reorganize this expression to avoid floating point erros, specially because of the common factor `1/m`

Reply:
I started doing this in 58857fd. Other than the common factor what are the possible issues? I think `sigma_internal`, `sigma_total_squared / m`, and `m` are all about the same order of magnitude but could be missing something.